File: androidEats/src/main/java/com/dunk/eats/android/presentation/recipe_browse/RecipeBrowseViewModel.kt
package com.dunk.eats.android.presentation.recipe_browse

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dunk.eats.domain.model.ErrorMessage
import com.dunk.eats.domain.model.Recipe
import com.dunk.eats.domain.model.UIComponentType
import com.dunk.eats.interactors.recipe_categories.Category
import com.dunk.eats.interactors.recipe_categories.CategoryTypes
import com.dunk.eats.interactors.recipe_list.SearchRecipes
import com.dunk.eats.presentation.recipe_browse.RecipeBrowseEvents
import com.dunk.eats.presentation.recipe_browse.RecipeBrowseState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.*
import javax.inject.Inject
import kotlin.collections.ArrayList

@HiltViewModel
class RecipeBrowseViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    private val searchRecipes: SearchRecipes,
    private val categoryTypes: CategoryTypes
) : ViewModel() {

    val state: MutableState<RecipeBrowseState> = mutableStateOf(RecipeBrowseState())

    init {
        onTriggerEvent(RecipeBrowseEvents.LoadCategories)
    }

    fun onTriggerEvent(event: RecipeBrowseEvents){
        when(event){
            RecipeBrowseEvents.LoadCategories -> {
                loadRecipes()
                loadPopularCategories()
            }

            RecipeBrowseEvents.NextPage -> {
                nextPage()
            }
            RecipeBrowseEvents.NewSearch -> {
                newSearch()
            }
            is RecipeBrowseEvents.OnUpdateQuery  -> {
                state.value = state.value.copy(query =  event.query)
            }
            is RecipeBrowseEvents.OnSelectCategory  -> {
                onSelectCategory(event.category)
            }
            else -> {
                addErrorToQueue(
                    ErrorMessage.Builder()
                        .id(UUID.randomUUID().toString())
                        .title("Error")
                        .uiComponentType(UIComponentType.Dialog)
                        .description("Invalid event")
                )
            }
        }
    }

    fun getFoodCategory(categoryName:String): Category? {
        return categoryTypes.getCategory(categoryName)
    }

    private fun onSelectCategory(category: Category) {
        state.value = state.value.copy(selectedCategory = category, query = category.categoryName)
        newSearch()
    }

    private fun newSearch() {
        state.value = state.value.copy(page = 1, recipes = listOf() )
        loadRecipes()
    }

    private fun nextPage() {
        //update the state and increment the page
        state.value = state.value.copy(page = state.value.page + 1)
        loadRecipes()
    }

    private fun loadRecipes(){
        searchRecipes.execute(
            page = state.value.page,
            query = state.value.query,
        ).onEach { dataState ->
            state.value= state.value.copy(isLoading = dataState.isLoading)

            dataState.data?.let { recipes ->
                appendRecipes(recipes)
            }

            dataState.errorMessage?.let { message ->
                addErrorToQueue(
                    ErrorMessage.Builder()
                        .id(UUID.randomUUID().toString())
                        .title(message.title)
                        .uiComponentType(UIComponentType.Dialog)
                        .description(message.description ?: "Unknown error")
                )
            }
        }.launchIn(viewModelScope)
    }

    private fun appendRecipes(recipes: List<Recipe>){
        val curr = ArrayList(state.value.recipes)
        curr.addAll(recipes)
        state.value = state.value.copy(recipes = curr)
    }

    private fun loadPopularCategories(){
        state.value = state.value.copy(categories = categoryTypes.getAllCategories())
    }

    private fun addErrorToQueue(error: ErrorMessage.Builder) {
        val queue = state.value.errorQueue
        queue.add(error.build())
        state.value = state.value.copy(errorQueue = queue)
    }

}

File: iosEats/iosEats/presentation/recipe_browse/RecipeBrowseViewModel.swift
//
//  RecipeBrowseViewModel.swift
//  iosEats
//
//  Created by Duncan K on 20/11/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import SwiftUI
import shared

class RecipeBrowseViewModel: ObservableObject {
    
    let searchRecipes: SearchRecipes
    let foodCategories: CategoryTypes
    var clickedRecipe: Recipe? = nil
    
    @Published var state: RecipeBrowseState = RecipeBrowseState()

    
    init(
        searchRecipes: SearchRecipes,
        foodCategories: CategoryTypes
    ){
        self.searchRecipes = searchRecipes
        self.foodCategories = foodCategories
        onTriggerEvent(stateEvent:RecipeBrowseEvents.LoadCategories())
    }
    
    func onTriggerEvent(stateEvent: RecipeBrowseEvents){
        switch stateEvent {
        case is RecipeBrowseEvents.LoadCategories :
            loadCategories()
        case is RecipeBrowseEvents.NextPage :
            loadNextPage()
        case is RecipeBrowseEvents.NewSearch :
            newSearch()
        case is RecipeBrowseEvents.OnUpdateQuery :
            onUpdateQuery(query: (stateEvent as!  RecipeBrowseEvents.OnUpdateQuery).query)
        case is RecipeBrowseEvents.OnSelectCategory :
            doNothing()
        case is RecipeBrowseEvents.LoadCategories :
            doNothing()
        case is RecipeBrowseEvents.RemoveHeadMessageFromQueue :
            doNothing()
        default:
            doNothing()
        }
    }
    
    func doNothing(){
        
    }
    
    func loadCategories(){
        let categories = foodCategories.getAllCategories()
        updateState(
            categories: categories
        )
    }
    
    func getTopCategories() -> [FoodCategories]{
        let currentState = (self.state.copy() as! RecipeBrowseState)
        return currentState.categories.filter{ $0.isPopular }
    }
    
    private func onUpdateQuery(query:String){
        updateState(query: query)
    }
    
    private func newSearch(){
        resetSearchState()
        loadRecipes()
    }
    
    private func resetSearchState(){
        let currentState = (self.state.copy() as! RecipeBrowseState)
        var foodCategory = currentState.selectedFoodCategories
        if(foodCategory?.name != currentState.query){
            foodCategory = nil
        }
        self.state = self.state.doCopy(
            isLoading: currentState.isLoading,
            page: 1,
            query: currentState.query,
            categories: currentState.categories,
            selectedFoodCategories: currentState.selectedFoodCategories,
            recipes: [],
            bottomRecipe: currentState.bottomRecipe,
            errorQueue: currentState.errorQueue
        )
    }
    
    private func loadRecipes() {
        let currentState = (self.state.copy() as! RecipeBrowseState)
        do{
            try searchRecipes.execute(
                page: Int32(currentState.page),
                query: currentState.query
            ).collectFlow(coroutineScope: nil, callback: { dataState in
                if dataState != nil {
                    let data = dataState?.data
                    let message = dataState?.errorMessage
                    let loading = dataState?.isLoading ?? false
                    
                    self.updateState(isLoading: loading)
                    
                    if data != nil {
                        self.appendRecipes(recipes: data as! [Recipe])
                    }
                    if message != nil {
                        self.displayErrorMessageByUIComponentType(message)
                    }
                }
            })
        }catch{
            print("\(error)")
        }
    }
    
    func appendRecipes(recipes: [Recipe]){
        var currentState = (self.state.copy()) as! RecipeBrowseState
        var currentRecipes = currentState.recipes
        currentRecipes.append(contentsOf: recipes)
        self.state = self.state.doCopy(
            isLoading: currentState.isLoading,
            page: currentState.page,
            query: currentState.query,
            categories: currentState.categories,
            selectedFoodCategories: currentState.selectedFoodCategories,
            recipes: currentRecipes ,
            bottomRecipe: currentState.bottomRecipe,
            errorQueue: currentState.errorQueue
        )
        
        currentState = self.state.copy() as! RecipeBrowseState
        self.updateBottomRecipe(recipe: currentState.recipes[currentState.recipes.count - 1])
    }
    
    func updateBottomRecipe(recipe:Recipe) {
        updateState(bottomRecipe: recipe)
    }
    
    func shouldLoadNextPage(recipe:Recipe) -> Bool {
        let currentState = (self.state.copy()) as! RecipeBrowseState
        if (recipe.id == currentState.bottomRecipe?.id) {
            if (SearchRecipes.Companion().RECIPE_PAGINATION_PAGE_SIZE *
                currentState.page <= currentState.recipes.count){
                if(!currentState.isLoading){
                    return true
                }
            }
        }
        return false
    }
    
    func loadNextPage(){
        let currentState = (self.state.copy() as! RecipeBrowseState)
        updateState(page: Int(currentState.page + 1))
        loadRecipes()
    }
    
    
    func updateState(
        isLoading: Bool? = nil,
        page:Int? = nil,
        query: String? = nil,
        categories: [FoodCategories]? = nil,
        selectedFoodCategory: FoodCategories? = nil,
        bottomRecipe: Recipe? = nil,
        queue: Queue<ErrorMessage>? = nil
    ){
        let currentState = (self.state.copy() as! RecipeBrowseState)
        
        self.state = self.state.doCopy(
            isLoading: isLoading ?? currentState.isLoading,
            page: Int32(page ?? Int(currentState.page)),
            query: query ?? currentState.query,
            categories: categories ?? currentState.categories,
            selectedFoodCategories: currentState.selectedFoodCategories,
            recipes: currentState.recipes,
            bottomRecipe: bottomRecipe ?? currentState.bottomRecipe,
            errorQueue: currentState.errorQueue
        )
    }
    
    func displayErrorMessageByUIComponentType( _ errorMessage: ErrorMessage?){
        //Todo handle errpr message
    }
    
}
